diff -up ./test/config.json.example.lindefault ./test/config.json.example
--- ./test/config.json.example.lindefault	2013-05-30 16:04:41.206533618 +0200
+++ ./test/config.json.example	2013-05-30 16:04:34.070449358 +0200
@@ -10,24 +10,24 @@
   // otherwise the tests will fail.
 
   //// Good values for OS X:
-  "normal_user_1": "_www",
-  "normal_user_2": "daemon",
-  // Must not be "nobody".
-  "default_user": "_sandbox",
-  // Must not be normal_user_1's primary group.
-  "normal_group_1": "daemon",
-  // Must not be normal_user_2's primary group.
-  "normal_group_2": "_sandbox",
-  // Must not be default_user's primary group. Must not be "nobody".
-  "default_group": "_www",
-
-  ///// Good values for Linux and FreeBSD. Same restrictions apply.
-  //"normal_user_1": "games",
+  //"normal_user_1": "_www",
   //"normal_user_2": "daemon",
-  //"default_user": "man",
+  //// Must not be "nobody".
+  //"default_user": "_sandbox",
+  //// Must not be normal_user_1's primary group.
   //"normal_group_1": "daemon",
-  //"normal_group_2": "man",
-  //"default_group": "games",
+  //// Must not be normal_user_2's primary group.
+  //"normal_group_2": "_sandbox",
+  //// Must not be default_user's primary group. Must not be "nobody".
+  //"default_group": "_www",
+
+  ///// Good values for Linux and FreeBSD. Same restrictions apply.
+  "normal_user_1": "games",
+  "normal_user_2": "daemon",
+  "default_user": "man",
+  "normal_group_1": "daemon",
+  "normal_group_2": "man",
+  "default_group": "games",
 
   // A nonexistant username, group name, user ID and group ID.
   "nonexistant_user": "xxxxxxxxxxxxxxxxxxx",
