diff -uNrbw passenger-3.0.0.pre4.upstream/bin/passenger-install-nginx-module passenger-3.0.0.pre4/bin/passenger-install-nginx-module
--- passenger-3.0.0.pre4.upstream/bin/passenger-install-nginx-module	2010-10-01 03:15:07.000000000 -0700
+++ passenger-3.0.0.pre4/bin/passenger-install-nginx-module	2010-10-18 16:32:06.000000000 -0700
@@ -92,7 +92,7 @@
 		check_whether_we_can_write_to(nginx_prefix) || exit(1)
 		nginx_config_already_exists_before_installing = nginx_config_exists?(nginx_prefix)
 		compile_passenger_support_files || exit(1)
-		if install_nginx(nginx_source_dir, nginx_prefix, extra_nginx_configure_flags)
+		if install_nginx(nginx_source_dir, nginx_prefix, extra_nginx_configure_flags, @extra_make_install_flags)
 			if nginx_config_already_exists_before_installing || !locate_nginx_config_file(nginx_prefix)
 				show_passenger_config_snippets(nginx_prefix)
 			else
@@ -322,13 +322,13 @@
 		return !!locate_nginx_config_file(prefix)
 	end
 	
-	def install_nginx(source_dir, prefix, extra_configure_flags)
+	def install_nginx(source_dir, prefix, extra_configure_flags, extra_make_install_flags)
 		Dir.chdir(source_dir) do
 			new_screen
 			color_puts "<banner>Compiling and installing Nginx...</banner>"
 			if !sh(build_nginx_configure_command(prefix, extra_configure_flags)) ||
 			   !sh("make") ||
-			   !sh("make install")
+			   !sh("make install #{extra_make_install_flags}")
 				return false
 			end
 		end
@@ -458,6 +458,11 @@
 	        "#{' ' * 37}installer to ask interactively either.") do |flags|
 		options[:extra_configure_flags] = flags
 	end
+	opts.on("--extra-make-install-flags=STRING", String, "Pass these extra flags to Ngin's\n" <<
+	        "#{' ' * 37}'make install' step. Only applicable if\n" <<
+	        "#{' ' * 37}--nginx-source-dir is given.") do |flags|
+    options[:extra_make_install_flags] = flags
+	end
 end
 begin
 	parser.parse!
