diff --git a/lib/phusion_passenger/standalone/utils.rb b/lib/phusion_passenger/standalone/utils.rb
index eceba5a..6f46932 100644
--- a/lib/phusion_passenger/standalone/utils.rb
+++ b/lib/phusion_passenger/standalone/utils.rb
@@ -33,7 +33,7 @@ private
 	end
 	
 	def runtime_version_string
-		if PhusionPassenger.natively_packaged?
+		if PhusionPassenger.natively_packaged? || File.exists?('/etc/redhat-release')
 			return "natively-packaged"
 		else
 			require_platform_info_binary_compatibility
