diff --git a/build/packaging.rb b/build/packaging.rb
index aa63b56..58044cd 100644
--- a/build/packaging.rb
+++ b/build/packaging.rb
@@ -108,10 +108,9 @@ task :fakeroot => [:apache2, :nginx] + Packaging::ASCII_DOCS do
 	fake_agents_dir = "#{fakeroot}#{NATIVELY_PACKAGED_AGENTS_DIR}"
 	fake_helper_scripts_dir = "#{fakeroot}#{NATIVELY_PACKAGED_HELPER_SCRIPTS_DIR}"
 	fake_resources_dir = "#{fakeroot}/usr/share/phusion-passenger"
-	fake_docdir = "#{fakeroot}#{NATIVELY_PACKAGED_DOCDIR}"
+	fake_docdir = "#{fakeroot}#{NATIVELY_PACKAGED_DOC_DIR}"
 	fake_bindir = "#{fakeroot}/usr/bin"
 	fake_sbindir = "#{fakeroot}/usr/sbin"
-	fake_source_root = "#{fakeroot}#{NATIVELY_PACKAGED_SOURCE_ROOT}"
 	fake_apache2_module = "#{fakeroot}#{NATIVELY_PACKAGED_APACHE2_MODULE}"
 	fake_apache2_module_dir = File.dirname(fake_apache2_module)
 	
@@ -119,24 +118,24 @@ task :fakeroot => [:apache2, :nginx] + Packaging::ASCII_DOCS do
 	sh "mkdir -p #{fakeroot}"
 	
 	sh "mkdir -p #{fake_libdir}"
-	sh "cp #{LIBDIR}/phusion_passenger.rb #{fake_libdir}/"
-	sh "cp -R #{LIBDIR}/phusion_passenger #{fake_libdir}/"
+	sh "cp #{PhusionPassenger.ruby_libdir}/phusion_passenger.rb #{fake_libdir}/"
+	sh "cp -R #{PhusionPassenger.ruby_libdir}/phusion_passenger #{fake_libdir}/"
 	
 	sh "mkdir -p #{fake_native_support_dir}"
-	native_support_archdir = PlatformInfo.ruby_extension_binary_compatibility_ids.join("-")
+	native_support_archdir = PlatformInfo.ruby_extension_binary_compatibility_id
 	sh "mkdir -p #{fake_native_support_dir}"
-	sh "cp -R ext/ruby/#{native_support_archdir}/*.#{LIBEXT} #{fake_native_support_dir}/"
+	sh "cp -R libout/ruby/#{native_support_archdir}/*.#{LIBEXT} #{fake_native_support_dir}/"
 	
 	sh "mkdir -p #{fake_agents_dir}"
-	sh "cp -R #{AGENTS_DIR}/* #{fake_agents_dir}/"
+	sh "cp -R #{PhusionPassenger.agents_dir}/* #{fake_agents_dir}/"
 	sh "rm -rf #{fake_agents_dir}/*.dSYM"
 	sh "rm -rf #{fake_agents_dir}/*/*.dSYM"
 	
 	sh "mkdir -p #{fake_helper_scripts_dir}"
-	sh "cp -R #{HELPER_SCRIPTS_DIR}/* #{fake_helper_scripts_dir}/"
+	sh "cp -R #{PhusionPassenger.helper_scripts_dir}/* #{fake_helper_scripts_dir}/"
 	
 	sh "mkdir -p #{fake_resources_dir}"
-	sh "cp resources/* #{fake_resources_dir}/"
+	sh "cp -R resources/* #{fake_resources_dir}/"
 	
 	sh "mkdir -p #{fake_docdir}"
 	Packaging::ASCII_DOCS.each do |docfile|
@@ -156,17 +155,8 @@ task :fakeroot => [:apache2, :nginx] + Packaging::ASCII_DOCS do
 	
 	sh "mkdir -p #{fake_apache2_module_dir}"
 	sh "cp #{APACHE2_MODULE} #{fake_apache2_module_dir}/"
-	
-	sh "mkdir -p #{fake_source_root}"
-	spec.files.each do |filename|
-		next if File.directory?(filename)
-		dirname = File.dirname(filename)
-		if !File.directory?("#{fake_source_root}/#{dirname}")
-			sh "mkdir -p '#{fake_source_root}/#{dirname}'"
-		end
-		puts "cp '#{filename}' '#{fake_source_root}/#{dirname}/'"
-		FileUtils.cp(filename, "#{fake_source_root}/#{dirname}/")
-	end
+
+	sh "find #{fakeroot} -name .DS_Store -print0 | xargs -0 rm -f"
 end
 
 desc "Create a Debian package"
