commit 4d3dbbc3ad08a9f1821e01e1c33e646470ef9755
Author: Hongli Lai (Phusion) <hongli@phusion.nl>
Date:   Wed May 22 18:07:42 2013 +0200

    Include headers in the fakeroot.

--- a/bin/passenger-config
+++ b/bin/passenger-config
@@ -36,6 +36,7 @@
 	puts
 	puts "Options:"
 	puts "  --root                     Show Phusion Passenger's root directory."
+	puts "  --includedir               Show the Nginx runtime library headers directory."
 	puts "  --nginx-libs               Show Nginx runtime library flags."
 	puts "  --compiled                 Check whether runtime libraries are compiled."
 	puts "  --ruby-command             Print the correct command for invoking the Ruby interpreter."
@@ -52,6 +53,8 @@
 case ARGV[0]
 when "--root"
 	puts PhusionPassenger.source_root
+when "--includedir"
+	puts PhusionPassenger.include_dir
 when "--nginx-libs"
 	text = "#{common_library.link_objects_as_string} #{PhusionPassenger.lib_dir}/common/libboost_oxt.a"
 	if PhusionPassenger::PlatformInfo.has_math_library?
--- a/build/packaging.rb
+++ b/build/packaging.rb
@@ -187,6 +187,7 @@
 	fake_agents_dir = "#{fakeroot}/usr/lib/#{GLOBAL_NAMESPACE_DIRNAME}/agents"
 	fake_helper_scripts_dir = "#{fakeroot}/usr/share/#{GLOBAL_NAMESPACE_DIRNAME}/helper-scripts"
 	fake_resources_dir = "#{fakeroot}/usr/share/phusion-passenger"
+	fake_include_dir = "#{fakeroot}/usr/share/phusion-passenger/include"
 	fake_docdir = "#{fakeroot}/usr/share/doc/#{GLOBAL_NAMESPACE_DIRNAME}"
 	fake_bindir = "#{fakeroot}/usr/bin"
 	fake_sbindir = "#{fakeroot}/usr/sbin"
@@ -221,6 +222,23 @@
 	
 	sh "mkdir -p #{fake_resources_dir}"
 	sh "cp -R resources/* #{fake_resources_dir}/"
+
+	sh "mkdir -p #{fake_include_dir}"
+	# Infer headers that the Nginx module needs
+	headers = []
+	Dir["ext/nginx/*.[ch]"].each do |filename|
+		File.read(filename).split("\n").grep(%r{#include "common/(.+)"}) do |match|
+			headers << ["ext/common/#{$1}", $1]
+		end
+	end
+	headers.each do |header|
+		target = "#{fake_include_dir}/#{header[1]}"
+		dir = File.dirname(target)
+		if !File.directory?(dir)
+			sh "mkdir -p #{dir}"
+		end
+		sh "cp #{header[0]} #{target}"
+	end
 	
 	sh "mkdir -p #{fake_docdir}"
 	Packaging::ASCII_DOCS.each do |docfile|
@@ -252,6 +270,7 @@
 		f.puts "libdir=/usr/lib/phusion-passenger"
 		f.puts "helper_scripts=/usr/share/phusion-passenger/helper-scripts"
 		f.puts "resources=/usr/share/phusion-passenger"
+		f.puts "includedir=/usr/share/phusion-passenger/include"
 		f.puts "doc=/usr/share/doc/phusion-passenger"
 		f.puts "rubylibdir=/usr/lib/ruby/vendor_ruby"
 		f.puts "apache2_module=/usr/lib/apache2/modules/mod_passenger.so"
--- a/doc/Packaging.txt.md
+++ b/doc/Packaging.txt.md
@@ -88,6 +88,7 @@
     libdir=/usr/lib/phusion-passenger
     helper_scripts=/usr/share/phusion-passenger/helper-scripts
     resources=/usr/share/phusion-passenger
+    includedir=/usr/share/phusion-passenger/include
     doc=/usr/share/doc/phusion-passenger
     rubylibdir=/usr/lib/ruby/vendor_ruby
     apache2_module=/usr/lib/apache2/modules/mod_passenger.so
@@ -193,6 +194,13 @@
 
    Value when originally packaged: `<SOURCE_ROOT>/doc`.
 
+ * `includedir`
+
+   A directory that contains the Phusion Passenger header files that are
+   necessary for compiling Nginx.
+
+   Value when originally packaged: `<SOURCE_ROOT>/ext`
+
  * `libdir`
 
    A directory that contains the Phusion Passenger library files, e.g.
--- a/ext/nginx/Configuration.c
+++ b/ext/nginx/Configuration.c
@@ -35,8 +35,8 @@
 #include "ngx_http_passenger_module.h"
 #include "Configuration.h"
 #include "ContentHandler.h"
-#include "../common/Constants.h"
-#include "../common/agents/LoggingAgent/FilterSupport.h"
+#include "common/Constants.h"
+#include "common/agents/LoggingAgent/FilterSupport.h"
 
 
 static ngx_str_t headers_to_hide[] = {
--- a/ext/nginx/ContentHandler.c
+++ b/ext/nginx/ContentHandler.c
@@ -31,7 +31,7 @@
 #include "ContentHandler.h"
 #include "StaticContentHandler.h"
 #include "Configuration.h"
-#include "../common/Constants.h"
+#include "common/Constants.h"
 
 
 #define NGX_HTTP_SCGI_PARSE_NO_HEADER  20
--- a/ext/nginx/ContentHandler.h
+++ b/ext/nginx/ContentHandler.h
@@ -30,7 +30,7 @@
 
 #include <ngx_core.h>
 #include <ngx_http.h>
-#include "../common/ApplicationPool2/AppTypes.h"
+#include "common/ApplicationPool2/AppTypes.h"
 
 
 typedef struct {
--- a/ext/nginx/config
+++ b/ext/nginx/config
@@ -33,8 +33,14 @@
     ${ngx_addon_dir}/ContentHandler.h \
     ${ngx_addon_dir}/StaticContentHandler.h \
     ${ngx_addon_dir}/ngx_http_passenger_module.h \
-    `passenger-config --root`/ext/common/Constants.h \
-    `passenger-config --root`/ext/common/ApplicationPool2/AppTypes.h"
+    `passenger-config --includedir`/common/Constants.h \
+    `passenger-config --includedir`/common/ApplicationPool2/AppTypes.h"
+if test "x$PASSENGER_INCLUDEDIR" = "x"; then
+    CORE_INCS="$CORE_INCS `passenger-config --includedir`"
+else
+    CORE_INCS="$CORE_INCS $PASSENGER_INCLUDEDIR"
+fi
+
 if test "x$PASSENGER_LIBS" = "x"; then
     CORE_LIBS="$CORE_LIBS `passenger-config --nginx-libs`"
 else
--- a/ext/nginx/ngx_http_passenger_module.h
+++ b/ext/nginx/ngx_http_passenger_module.h
@@ -30,9 +30,9 @@
 
 #include <ngx_config.h>
 #include <ngx_core.h>
-#include "../common/AgentsStarter.h"
-#include "../common/ApplicationPool2/AppTypes.h"
-#include "../common/Utils/CachedFileStat.h"
+#include "common/AgentsStarter.h"
+#include "common/ApplicationPool2/AppTypes.h"
+#include "common/Utils/CachedFileStat.h"
 
 /**
  * The Nginx version number as an integer.
--- a/lib/phusion_passenger.rb
+++ b/lib/phusion_passenger.rb
@@ -87,6 +87,7 @@
 			@lib_dir               = get_option(filename, options, 'libdir').freeze
 			@helper_scripts_dir    = get_option(filename, options, 'helper_scripts').freeze
 			@resources_dir         = get_option(filename, options, 'resources').freeze
+			@include_dir           = get_option(filename, options, 'includedir').freeze
 			@doc_dir               = get_option(filename, options, 'doc').freeze
 			@apache2_module_path   = get_option(filename, options, 'apache2_module').freeze
 			@ruby_extension_source_dir = get_option(filename, options, 'ruby_extension_source').freeze
@@ -98,6 +99,7 @@
 			@lib_dir               = "#{@source_root}/libout".freeze
 			@helper_scripts_dir    = "#{@source_root}/helper-scripts".freeze
 			@resources_dir         = "#{@source_root}/resources".freeze
+			@include_dir           = "#{@source_root}/ext".freeze
 			@doc_dir               = "#{@source_root}/doc".freeze
 			@apache2_module_path   = "#{@source_root}/libout/apache2/mod_passenger.so".freeze
 			@ruby_extension_source_dir = "#{@source_root}/ext/ruby"
@@ -139,6 +141,10 @@
 	def self.resources_dir
 		return @resources_dir
 	end
+
+	def self.include_dir
+		return @include_dir
+	end
 	
 	def self.doc_dir
 		return @doc_dir
--- a/lib/phusion_passenger/standalone/command.rb
+++ b/lib/phusion_passenger/standalone/command.rb
@@ -190,6 +190,7 @@
 			else
 				f.puts "agents=#{@runtime_dirs[:support_dir]}"
 			end
+			f.puts "libdir=#{PhusionPassenger.lib_dir}"
 			f.puts "helper_scripts=#{PhusionPassenger.helper_scripts_dir}"
 			f.puts "resources=#{PhusionPassenger.resources_dir}"
 			f.puts "doc=#{PhusionPassenger.doc_dir}"
--- a/lib/phusion_passenger/standalone/runtime_installer.rb
+++ b/lib/phusion_passenger/standalone/runtime_installer.rb
@@ -529,7 +529,8 @@
 				nginx_libs = COMMON_LIBRARY.only(*NGINX_LIBS_SELECTOR).
 					set_output_dir(output_dir).
 					link_objects_as_string
-				command << "env PASSENGER_LIBS='#{nginx_libs} #{output_dir}/../libboost_oxt.a' "
+				command << "env PASSENGER_INCLUDEDIR='#{PhusionPassenger.include_dir}'" <<
+					" PASSENGER_LIBS='#{nginx_libs} #{output_dir}/../libboost_oxt.a' "
 			end
 			# RPM thinks it's being smart by scanning binaries for
 			# paths and refusing to create package if it detects any
